esmvalcore/cmor/_fixes/common.py

```diff
@@ -108,49 +109,70 @@ def fix_metadata(self, cubes):
 class OceanFixGrid(Fix):
     """Fixes for tos, siconc in FGOALS-g3."""
 
-    def fix_data(self, cube):
-        """
-        Fix data.
-
-        Calculate missing latitude/longitude boundaries using interpolation.
-        Based on a similar fix for BCC-CSM2-MR.
+    def fix_metadata(self, cubes):
+        """Fix ``latitude`` and ``longitude`` (metadata and bounds).
 
         Parameters
         ----------
-        cube: iris.cube.Cube
-            Input cube to fix.
+        cubes : iris.cube.CubeList
+            Input cubes.
 
         Returns
         -------
-        iris.cube.Cube
+        iris.cube.CubeList
+
         """
-        rlat = cube.coord('grid_latitude').points
-        rlon = cube.coord('grid_longitude').points
-
-        # Guess coordinate bounds in rlat, rlon (following BCC-CSM2-MR-1).
-        rlat_idx_bnds = np.zeros((len(rlat), 2))
-        rlat_idx_bnds[:, 0] = np.arange(len(rlat)) - 0.5
-        rlat_idx_bnds[:, 1] = np.arange(len(rlat)) + 0.5
-        rlat_idx_bnds[0, 0] = 0.
-        rlat_idx_bnds[len(rlat) - 1, 1] = len(rlat)
-        rlon_idx_bnds = np.zeros((len(rlon), 2))
-        rlon_idx_bnds[:, 0] = np.arange(len(rlon)) - 0.5
-        rlon_idx_bnds[:, 1] = np.arange(len(rlon)) + 0.5
+        cube = self.get_cube_from_list(cubes)
+
+        # Get dimensional coordinates. Note:
+        # - First dimension i -> X-direction (longitude; = dimension 2 in data)
+        # - Second dimension j -> Y-direction (latitude; = dimension 1 in data)
+        i_coord = cube.coord(dim_coords=True, dimensions=2)
+        j_coord = cube.coord(dim_coords=True, dimensions=1)
```

Comment:
Yeah, it's also not my preferred solution, but there are 3 different names for this coordinate among the different models: For it's CMIP5 model, BCC uses `grid_latitude`; for it's CMIP6 model, BCC uses `latitude`; and FGOALS uses `j`. So referencing them by name is not an option.

However, I don't see an issue here regarding 3D variables, since the fix is only applied to 2D variables (and there is no automatic application of it). If we want to support 3D variables, the fix needs to be expanded accordingly, but this is not in the scope of this PR.

Reply:
Regarding `i` and `j`: I just followed FGOALS here, who use `i` for their X- and `j` for their Y-coordinate. I think it's also consistent with the [CMOR table](https://github.com/ESMValGroup/ESMValCore/blob/fee5d541dc871efe72ab972a6a52195c452d16eb/esmvalcore/cmor/tables/cmip6/Tables/CMIP6_grids.json#L35-L49) that defines `i` as `first spatial index for variables stored on an unstructured grid` (which sounds like X to me) and `j` as `second spatial index for variables stored on an unstructured grid` (which sounds like Y to me).

Reply:
Looking at various models, I found both conventions in use. If we look at the data in our environment, ie look at, eg `tos[k, l]` I think the first index varies latitudes, the second longitudes. I will try to escalate this upstream for clarification in CMIP7.

Reply:
Setting aside the issue of i and j, I think we can probably figure out a way to identify the dimensions which makes it possible to apply the fix also to 3d data. In my view that would be highly desirable because the models that suffer from this problem in 2d ocean will have the same problem in 3d ocean data (e.g. (Omon, thetao), used for heat content analysis).

How about
```
y_dim, x_dim = sorted(set(cube.coord_dims('latitude') + cube.coord_dims('longitude')))
```
?

Reply:
I agree that writing more general fixes is generally a good idea, but I still think that this is out of scope for this PR. Right now, the fix is only applied to a set of predefined variables, which are all 2D. If we want to include more variables and also 3D data, we should do this properly in a separate PR.

The following things might be relevant for a corresponding PR:

- If we think that all variables on ocean grids are affected, we should put this fix into an `AllVars` class and include a proper check for the presence of an ocean grid (maybe if lat and lon are 2D?).
- If we want to support 3D data, we need to include units tests for this and test it on real data.

Given that these are non-trivial changes to the current form of the fix, I think we shouldn't address this here.

Reply:
I agree that we should have a separate PR to include this in allvars and to add tests for 3d. Nevertheless, this should never have been written with constant dimensions and the improvement to the present code is simple enough to include here.

Reply:
This does not work for BCC's CMIP6 models, since the 1D coords are also called `latitude` and `longitude`:

```
iris.exceptions.CoordinateNotFoundError: 'Expected to find exactly 1 coordinate, but found 2. They were: latitude, latitude.'
```

Reply:
Okay, I found a way around it and implemented it. I still don't like the idea of including code that is supposed to work for more general cases without testing it on these cases once.

Reply:
Noted. Do you want to emit a warning if the dimensionality is too high?

Reply:
Done